File: app/src/main/java/com/glucocontrol/domain/usecase/reading/UpdateReadingUseCase.kt
package com.glucocontrol.domain.usecase.reading

import com.glucocontrol.domain.model.GlucoseReading
import com.glucocontrol.domain.repository.GlucoseRepository
import java.time.LocalDate
import javax.inject.Inject

/**
 * Actualiza todos los campos de una lectura de glucosa existente.
 *
 * Aplica las mismas validaciones de negocio que [AddReadingUseCase] más
 * la comprobación de que el id sea válido (> 0), garantizando que el
 * registro existe antes de intentar modificarlo.
 */
class UpdateReadingUseCase
@Inject
constructor(private val repository: GlucoseRepository) {
    private val minimumDate = LocalDate.of(2020, 1, 1)

    suspend operator fun invoke(reading: GlucoseReading) {
        require(reading.id > 0) {
            "No se puede actualizar un registro sin id válido (id = ${reading.id})"
        }
        require(reading.valueMgDl > 0) {
            "El valor de glucosa debe ser mayor que cero"
        }
        require(reading.valueMgDl <= 600) {
            "El valor de glucosa supera el máximo admitido (600 mg/dL)"
        }
        require(!reading.date.isBefore(minimumDate)) {
            "No se admiten registros anteriores al 1 de enero de 2020"
        }
        repository.updateReading(reading)
    }
}

File: lib/domain/usecase/reading/update_reading_use_case.dart
import 'package:glucocontrol/domain/model/glucose_reading.dart';
import 'package:glucocontrol/domain/repository/glucose_repository.dart';

class UpdateReadingUseCase {
  final GlucoseRepository _repo;
  UpdateReadingUseCase(this._repo);

  static final _minDate = DateTime(2020, 1, 1);

  Future<void> call(GlucoseReading reading) {
    if (reading.id <= 0) throw ArgumentError('ID inválido');
    if (reading.valueMgDl < 1 || reading.valueMgDl > 600) {
      throw ArgumentError('La glucosa debe estar entre 1 y 600 mg/dL');
    }
    final date = DateTime(reading.date.year, reading.date.month, reading.date.day);
    if (date.isBefore(_minDate)) {
      throw ArgumentError('La fecha no puede ser anterior a 2020-01-01');
    }
    return _repo.updateReading(reading);
  }
}
